core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -91,19 +91,24 @@ class DelayedFetch(
             // Go directly to the check for Case G if the message offsets are the same. If the log segment
             // has just rolled, then the high watermark offset will remain the same but be on the old segment,
             // which would incorrectly be seen as an instance of Case F.
-            if (endOffset.messageOffset != fetchOffset.messageOffset) {
-              if (endOffset.onOlderSegment(fetchOffset)) {
-                // Case F, this can happen when the new fetch operation is on a truncated leader
-                debug(s"Satisfying fetch $this since it is fetching later segments of partition $topicIdPartition.")
-                return forceComplete()
+            if (fetchOffset.messageOffset > endOffset.messageOffset) {
+              // Case F, this can happen when the new fetch operation is on a truncated leader
+              debug(s"Satisfying fetch $this since it is fetching later segments of partition $topicIdPartition.")
+              return forceComplete()
+            } else if (fetchOffset.messageOffset < endOffset.messageOffset) {
+              if (endOffset.messageOffsetOnly() || fetchOffset.messageOffsetOnly()) {
+                // If we don't know the position of the offset on log segments, just pessimistically assume that we
+                // only gained 1 byte when fetchOffset < endOffset, otherwise do nothing. This can happen when the
+                // high-watermark is stale, but should be rare.
+                accumulatedSize += 1
```

Comment:
If endOffset or fetchOffset is message only, we return empty records when reading from the log. So, to be consistent, we want to avoid accumulating new bytes in this case.

Reply:
Yes, in that case, we just wait until endOffset moves to within the local log segments. If the timeout hits, we will just return empty.